### ISSUE TITLE
Warn on &mut to a temporary value

### ISSUE THREAD

**Issue by redbaron** [rust-lang/rust-clippy#8712]
### What it does

Warn when &mut is taken on a temporary value, especially when type implements `Copy`. Code below doesn't modify `buf`, because what is being passed to `t` is a `&mut` to a temporary array created by a `<[u8]>::try_into()`:

`&mut` reference implies that a receiver will modify value, modifying temporary value, which user has no access to is pointless IMHO. More likely than not, it is not user intention.

### Lint Name

mut-ref-rvalue

### Category

suspicious

### Example

```rust
#[inline(never)]
pub fn t(b: &mut [u8;4]) {
    b[0] = 99;
}

#[inline(never)]
pub fn tt() -> [u8;16] {
    let buf = [0u8; 16];
    t(&mut buf[..4].try_into().unwrap());
    buf
}
```

Could be written as:

```rust
#[inline(never)]
pub fn t(b: &mut [u8;4]) {
    b[0] = 99;
}

#[inline(never)]
pub fn tt() -> [u8;16] {
    let mut buf = [0u8; 16];
    t((&mut buf[..4]).try_into().unwrap());
    buf
}
```

**Comment by Jarcho** [comment]
`suspicious` is a little too strong for this in general. `(&mut *x).write(buf)` is definitely used. Limited to functions which only take a single `&mut` parameter without a return value would be reasonable for suspicious, but more than that and you risk running into a real false positive rate.

**Comment by pitaj** [comment]
It would be hard to exclude all of the false positive cases like `(&mut *x).foo()` etc.

Maybe this should instead cover a known number of specific troublesome cases or maybe they should be implemented as separate lints.
